### ISSUE TITLE
Request working GPEN for AMD DirectX12 GPU

### ISSUE THREAD

**Issue by JohnC05** [yangxy/GPEN#120]
Hello,

I don't have cuda enabled device and currently using amd 5700xt gpu.
GPEN cpu version working fine but slow around 10~15 seconds per face, takes minutes for 4k image with multiple faces.
So i would like to speed up the process for large number of images with many faces by using amd directx12 gpu.

I have tried many times past 2 months to convert device='cuda' to device='dml' by using latest pytorch-directml with no luck.
Most of the time stuck at 'retinaface_detection.py' -> line 78 'loc, conf, landms = self.net(img)  # forward pass' debugging with dml.
There are also possibility with ncnn-vulkan.

So, i would like to request adding some sort of usability for amd directx12 gpu with GPEN out of the box.

Thank You.

N.B. also if any kind of tips to use it on amd gpu anyhow is much appreciated.

**Comment by yangxy** [comment]
I am not a user of amd directx12 gpu. Maybe I cannot help you with that. 
"Most of the time stuck at 'retinaface_detection.py' -> line 78 'loc, conf, landms = self.net(img) ", if so, you can replace the retinaface detection with other alternatives, such as [yolov5face](https://github.com/deepcam-cn/yolov5-face) and so on.

**Comment by JohnC05** [comment]
> I am not a user of amd directx12 gpu. Maybe I cannot help you with that. "Most of the time stuck at 'retinaface_detection.py' -> line 78 'loc, conf, landms = self.net(img) ", if so, you can replace the retinaface detection with other alternatives, such as [yolov5face](https://github.com/deepcam-cn/yolov5-face) and so on.

Thank you for the suggestion. I was beginning to try other facedetect model as you suggested but somehow i have managed it to work with your current gpen facedetector on pytorch-directml.
1024x1024p sized file now detects face just under **250ms (amd gpu)** instead of **4000ms (amd cpu)**, wonderfully.

But right after that it throws error at _"img_t = torch.from_numpy(img).to(self.device)/255."_ in img2tensor inside file face_gan.py

(solved) Error: RuntimeError: CHECK failed: ((((HRESULT)((backend_->dml_device->CompileOperator(op.Get(), execution_flags, __uuidof(**(&compiled_op_)), IID_PPV_ARGS_Helper(&compiled_op_))))) >= 0)) == (true):

So, i have changed it to _"img_t = torch.div((torch.from_numpy(img)).type(torch.float32), 255).to(self.device)"_ here self.device='dml'
I don't know if it is the right conversion of "img_t" or not but it now works and continue to run until it gets stuck at "inputs = ecd(inputs)" in class FullGenerator(nn.Module) inside file gpen_model.py

(stuck) Error: [F D:\a\_work\1\s\pytorch-directml\caffe2\dml\dml_tensor_desc.cc:81] Check failed: rank <= DML_TENSOR_DIMENSION_COUNT_MAX

It was throwing this one line error without any trails how to get there but i found "inputs = ecd(inputs)" triggers it.
after ecd0 next ecd1 doesn't takes inputs and shows error and breaks. But in cpu mode it works fine with all these changes.
I have no clue whether it is dimension of gpen model or the image dimension.

With facedetect success i am hoping someone can help me solve this and continue this journey. :)

**Comment by JohnC05** [comment]
It is done! managed to use ONNX and now it runs super fast, almost 60 times faster on my directx12 GPU.
Total 100 images. Resolution: 4K.

CPU ( AMD Ryzen 3900x ) : 120 minutes.

GPU ( AMD 5700XT )         :      2 minutes.

**Comment by marshalleq** [comment]
Hi this sounds great - would you mind elaborating your last step?  What did you do to get it to work with ONNX?  I'm not a developer so it takes me a bit longer to figure it all out.  Thanks.

**Comment by JohnC05** [comment]
> Hi this sounds great - would you mind elaborating your last step? What did you do to get it to work with ONNX? I'm not a developer so it takes me a bit longer to figure it all out. Thanks.

Hello, thank you for your interest. I will be glad to help.
Although i am also no expert and it was long time ago,
But i will try my best to recall those processes.

**How to run GPEN on ONNX**
There are 3 steps to complete the process.
1.   Convert pytorch model to onnx model.
2.   Load onnx model.
3.   Use the onnx model.

But first of all install required packages inside your python environment. Such as, i have used directml so my exact packages are onnx, onnxruntime-directml, torch-directml, 

**Convert pytorch model to onnx model:** Use the onnx_export.py from gpen misc folder to convert the model from _**.pth**_ to _**.onnx**_ .

In onnx_export.py dummy input needs to be changed according to match the model. At line 36 and 94,
`dummy_input = Variable(torch.randn(1, 3, 256, 256)) # for model GPEN-BFR-256` or
`dummy_input = Variable(torch.randn(1, 3, 512, 512)) # for model GPEN-BFR-512`

Also change the model inputs according to the model. At line 106,
`model = FullGenerator(256, 512, 8, 1, narrow=0.5) # for model GPEN-BFR-256 `or
`model = FullGenerator(512, 512, 8, 2, narrow=1) # for model GPEN-BFR-512`

Follow the argument parser and use the command as needed.
Example command: _**python onnx_export.py --model GPEN-BFR-256 --path "weights" --export**_
It should now produce the GPEN-BFR-256.onnx file inside the weights folder. If not working then add command --force-cpu.

**Load ONNX model:** Now that we have .onnx model file, we have to load it instead of .pth model file.

We have to edit the file inside GPEN folder face_model/face_gan.py at line 17 change
`self.mfile = os.path.join(base_dir, 'weights', model+'.pth')` to 
`self.onnxfile = os.path.join(base_dir, 'weights', model+'.onnx')`

At line 26 load_model should have only 3 lines of code,
```
    def load_model(self, channel_multiplier=2, narrow=1):
          so = onnxruntime.SessionOptions()
          so.log_severity_level = 3
          self.ort_session = onnxruntime.InferenceSession(self.onnxfile, so, providers=["DmlExecutionProvider"])
```

For my AMD device i have used directml but for your device you should check which ExecutionProvider is needed from https://onnxruntime.ai/docs/execution-providers/

Now the onnx model is loaded on the desired device and ready to use.

**Use the ONNX model:** In the same file face_gan.py at line 37
Change these two lines of code
```
with torch.no_grad():
            out, __ = self.model(img_t)
```
to
```
def to_numpy(tensor):
            return tensor.detach().cpu().numpy() if tensor.requires_grad else tensor.cpu().numpy()
ort_inputs = {self.ort_session.get_inputs()[0].name: to_numpy(img_t) }
ort_outs = self.ort_session.run(None, ort_inputs)
out = torch.tensor(ort_outs[0]).cpu()
```

And that's it. Run GPEN as usual and it should now work on ONNX runtime with maximum speed :)
I hope this helps, if any problem arise please don't hesitate to ask. Good Luck.

**Comment by marshalleq** [comment]
Wow!  Thanks so much, I can't wait to try this!  Will post back when I have given it a go!  Thanks again!

**Comment by marshalleq** [comment]
> > Hi this sounds great - would you mind elaborating your last step? What did you do to get it to work with ONNX? I'm not a developer so it takes me a bit longer to figure it all out. Thanks.
> 
> Hello, thank you for your interest. I will be glad to help. Although i am also no expert and it was long time ago, But i will try my best to recall those processes.
> 
> **How to run GPEN on ONNX** There are 3 steps to complete the process.
> 
>     1. Convert pytorch model to onnx model.
> 
>     2. Load onnx model.
> 
>     3. Use the onnx model.
> 
> 
> But first of all install required packages inside your python environment. Such as, i have used directml so my exact packages are onnx, onnxruntime-directml, torch-directml,
> 
> **Convert pytorch model to onnx model:** Use the onnx_export.py from gpen misc folder to convert the model from _**.pth**_ to _**.onnx**_ .
> 
> In onnx_export.py dummy input needs to be changed according to match the model. At line 36 and 94, `dummy_input = Variable(torch.randn(1, 3, 256, 256)) # for model GPEN-BFR-256` or `dummy_input = Variable(torch.randn(1, 3, 512, 512)) # for model GPEN-BFR-512`
> 
> Also change the model inputs according to the model. At line 106, `model = FullGenerator(256, 512, 8, 1, narrow=0.5) # for model GPEN-BFR-256 `or `model = FullGenerator(512, 512, 8, 2, narrow=1) # for model GPEN-BFR-512`
> 
> Follow the argument parser and use the command as needed. Example command: _**python onnx_export.py --model GPEN-BFR-256 --path "weights" --export**_ It should now produce the GPEN-BFR-256.onnx file inside the weights folder. If not working then add command --force-cpu.
> 
> **Load ONNX model:** Now that we have .onnx model file, we have to load it instead of .pth model file.
> 
> We have to edit the file inside GPEN folder face_model/face_gan.py at line 17 change `self.mfile = os.path.join(base_dir, 'weights', model+'.pth')` to `self.onnxfile = os.path.join(base_dir, 'weights', model+'.onnx')`
> 
> At line 26 load_model should have only 3 lines of code,
> 
> ```
>     def load_model(self, channel_multiplier=2, narrow=1):
>           so = onnxruntime.SessionOptions()
>           so.log_severity_level = 3
>           self.ort_session = onnxruntime.InferenceSession(self.onnxfile, so, providers=["DmlExecutionProvider"])
> ```
> 
> For my AMD device i have used directml but for your device you should check which ExecutionProvider is needed from https://onnxruntime.ai/docs/execution-providers/
> 
> Now the onnx model is loaded on the desired device and ready to use.
> 
> **Use the ONNX model:** In the same file face_gan.py at line 37 Change these two lines of code
> 
> ```
> with torch.no_grad():
>             out, __ = self.model(img_t)
> ```
> 
> to
> 
> ```
> def to_numpy(tensor):
>             return tensor.detach().cpu().numpy() if tensor.requires_grad else tensor.cpu().numpy()
> ort_inputs = {self.ort_session.get_inputs()[0].name: to_numpy(img_t) }
> ort_outs = self.ort_session.run(None, ort_inputs)
> out = torch.tensor(ort_outs[0]).cpu()
> ```
> 
> And that's it. Run GPEN as usual and it should now work on ONNX runtime with maximum speed :) I hope this helps, if any problem arise please don't hesitate to ask. Good Luck.

Hi, so I've got the cpu version running and am working my way through your AMD instructions.  Did you come across the issue where  onnxruntime-directml won't install?  onnxruntime does install, so does torch-directml, just not onnxruntime-directml.  I've tried on both python 3.9 and 3.10, both of which are supported.  

I keep getting the below.
pip3 install onnxruntime-directml
ERROR: Could not find a version that satisfies the requirement onnxruntime-directml (from versions: none)
ERROR: No matching distribution found for onnxruntime-directml

pip is the latest version and clearly both of my python versions are supported, so I'm wondering if you came across the same issue and solved it somehow?

Thanks!

**Comment by marshalleq** [comment]
So apparently onnxruntime-directml is only supported in linux, if linux is running under Windows Subsystem for Linux.  I did wonder if it was windows only, but not this kind of hybrid scenario.  @JohnC05 did you install this under windows, or is the package name a mistake?  I remember you sort of put this together for me from memory.  I will try running it without anyway.  Thanks.

**Comment by JohnC05** [comment]
> 

I am using this in my windows 10 operating system and for python 3.10.x, miniconda virtual environment with all the latest directml related packages that I mentioned from pypi.

Your error is most probably because of different software distribution repo. Are you using conda forge?

**Comment by marshalleq** [comment]
I'm on bare metal ubuntu.  I hate to say it but I just went out and got an nvidia card based on the response from Microsoft.

**Comment by JohnC05** [comment]
> I'm on bare metal ubuntu. I hate to say it but I just went out and got an nvidia card based on the response from Microsoft. I could have run this on Windows, I actually thought it didn't run on Windows at all, so that's my mistake. Now I have two GPU's!

Good for you, now you can run any app hassle free 😀 I wish I could but if it resolved by coding then I just can't. 😄

The main concern here was running GPEN on AMD gpu which supports direct x 12.
Directx is Microsoft product. So it was obvious that the whole process was meant for windows. Directml is the only way to achieve that goal.

For Linux based os there is a way for AMD is called ROCM. Which is probably faster. I didn't try yet.

**Comment by marshalleq** [comment]
Yes, I'm pretty new to ML so haven't gotten the lay of the land yet.  Expensive mistake getting the AMD GPU, it was a rush buy - which I never do, but when I saw the price of NVIDIA I figured I could work it all out.  But should have realised!

Not sure how much ROCM is out there yet.

The silly thing is that machine does dual boot windows, so if I had realised I would have done that.  But now like you say I can try all sorts of things that I couldn't try before.  So I'm happy.  And the Nvidia card does run a lot cooler - like about half!  I love AMD CPU's, I'm not sure about their GPU's yet.  Probably worthwhile for gaming, ML is a growing area though.
